x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/RateLimiter.java

```diff
@@ -0,0 +1,144 @@
+        Objects.requireNonNull(newUnit);
+
+        if (newAccumulatedTokensLimit < 0) {
+            throw new IllegalArgumentException("Accumulated tokens limit must be greater than or equal to 0");
+        }
+
+        if (newAccumulatedTokensLimit == Double.POSITIVE_INFINITY) {
+            throw new IllegalArgumentException(
+                Strings.format("Accumulated tokens limit must be less than or equal to %s", Double.MAX_VALUE)
+            );
+        }
+
+        if (newTokensPerTimeUnit <= 0) {
+            throw new IllegalArgumentException("Tokens per time unit must be greater than 0");
+        }
+
+        if (newTokensPerTimeUnit == Double.POSITIVE_INFINITY) {
+            throw new IllegalArgumentException(Strings.format("Tokens per time unit must be less than or equal to %s", Double.MAX_VALUE));
+        }
+
+        accumulatedTokens = Math.min(accumulatedTokens, newAccumulatedTokensLimit);
+
+        accumulatedTokensLimit = newAccumulatedTokensLimit;
+
+        var unitsInNanos = newUnit.toNanos(1);
+        tokensPerNanos = newTokensPerTimeUnit / unitsInNanos;
+        assert tokensPerNanos != Double.POSITIVE_INFINITY : "Tokens per nanosecond should not be infinity";
+
+        accumulateTokens();
+    }
+
+    /**
+     * Causes the thread to wait until the tokens are available
+     * @param tokens the number of items of work that should be throttled, typically you'd pass a value of 1 here
+     * @throws InterruptedException
+     */
+    public void acquire(int tokens) throws InterruptedException {
+        if (tokens <= 0) {
+            throw new IllegalArgumentException("Requested tokens must be positive");
+        }
+
+        double nanosToWait;
+        synchronized (this) {
+            accumulateTokens();
+            var accumulatedTokensToUse = Math.min(tokens, accumulatedTokens);
+            var additionalTokensRequired = tokens - accumulatedTokensToUse;
+            nanosToWait = additionalTokensRequired / tokensPerNanos;
+            accumulatedTokens -= accumulatedTokensToUse;
+            nextTokenAvailability = nextTokenAvailability.plus(Duration.ofNanos((long) nanosToWait));
+        }
+
+        sleeper.sleep((long) nanosToWait);
+    }
+
+    private void accumulateTokens() {
+        var now = Instant.now(clock);
+        if (now.isAfter(nextTokenAvailability)) {
+            var elapsedTimeNanos = nanosBetweenExact(nextTokenAvailability, now);
+            var newTokens = tokensPerNanos * elapsedTimeNanos;
+            accumulatedTokens = Math.min(accumulatedTokensLimit, newTokens);
+            nextTokenAvailability = now;
+        }
+    }
+
+    private static double nanosBetweenExact(Instant start, Instant end) {
+        try {
+            return ChronoUnit.NANOS.between(start, end);
+        } catch (ArithmeticException e) {
+            if (end.isAfter(start)) {
+                return Double.POSITIVE_INFINITY;
+            }
+
+            return Double.NEGATIVE_INFINITY;
```

Comment:
In the accumulateTokens code:
```
            var newTokens = tokensPerNanos * elapsedTimeNanos;
            accumulatedTokens = Math.min(accumulatedTokensLimit, newTokens);
```

`Math.min(a_positive_number, Double.NEGATIVE_INFINITY)` returns `Double.NEGATIVE_INFINITY`. If `accumulatedTokens` becomes a -ve number I think that could cause errors.

One option is to return 0 ( not +ve or -ve infinity).  Using ChronoUnit.MICRO or ChronoUnit.MILLIS reduces the chance of an arithmetic overflow

Reply:
Good point. I'll switch it to 0 and use micros instead.